[Bug]: (viz) Selenium & Webdriver Manager interdependency updates, altair-saver incompatible with altair 5, WebDriver dangling resource

### Describe the Bugs

1. `altair_saver` is [not fully compatible](https://altair-viz.github.io/user_guide/saving_charts.html#altair-saver) with Altair 5.0 API, it is recommended to use [vl-convert-python](https://altair-viz.github.io/user_guide/saving_charts.html#vl-convert).
2. Breaking changes introduced to [Chrome WebDriver](https://developer.chrome.com/docs/chromium/new-headless), multiple updates to combination of Selenium+WebDriver Manager. Support for older Chrome WebDrivers are being deprecated and will soon creep into cache-refreshed/ new installs.
3. When `nesta_ds_utils.viz.altair.saving.py: save()` is called without passing in a WebDriver object, `_google_chrome_driver_setup()` is called to instantiate a Chrome WebDriver, which is left running in the background with dangling resources/ refs.

The above two bugs are intermingled with wider API revamps in the respective dependencies. It is worthwhile to introduce these at `nesta_ds_utils`, then adapt to bugs discovered downstream (and review downstream 'product' maintenance).

### Instructions to Reproduce the Bug

1. Install `[dev]` extra, then `pip install --upgrade altair`, optionally `pip install --upgrade altair-viewer altair-saver`, then run `pytest`.
2. Same as above.
3. Change Selenium/WebDriverManager/Altair dependency versions with `pip install --upgrade dependency==another.version`, re-run `pytest`.

### Traceback from pytest
```
tests/cleaning/test_datetime_ops.py ..........                                                                                                                      [ 15%]
tests/loading_saving/test_S3.py ...............................                                                                                                     [ 64%]
tests/loading_saving/test_file_ops.py .....                                                                                                                         [ 71%]
tests/networks/test_build.py .............                                                                                                                          [ 92%]
tests/viz/altair/test_format.py .                                                                                                                                   [ 93%]
tests/viz/altair/test_saving.py F.F.                                                                                                                                [100%]

================================================================================ FAILURES =================================================================================
__________________________________________________________________________ test_save_png_altair ___________________________________________________________________________

    def test_save_png_altair():
        """Test that that figures saved as png exist."""
        fig = alt.Chart(pd.DataFrame()).mark_bar()
        path = "tests/temp/"
>       saving.save(fig, "test_fig", path)

tests/viz/altair/test_saving.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:120: in save
    _save_png(fig, path, name, scale_factor, driver)
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:43: in _save_png
    alt_saver.save(
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_saver/_core.py:172: in save
    return saver.save(fp=fp, fmt=fmt)
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_saver/savers/_saver.py:121: in save
    content = self._serialize(fmt, "save")
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_saver/savers/_selenium.py:284: in _serialize
    out = self._extract(fmt)
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_saver/savers/_selenium.py:237: in _extract
    "vega-lite.js": get_bundled_script(
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_viewer/_scripts.py:40: in get_bundled_script
    version_str = find_version(version, listing[package])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

version = '5.16.3', candidates = ['4.0.2', '4.8.1', '4.17.0'], strict_micro = False

    def find_version(
        version: Optional[str], candidates: List[str], strict_micro: bool = False
    ) -> str:
        """Find a matching version string given a list of candidate versions.
    
        Parameters
        ----------
        version : str or None
            The version to match. If None, the newest non-development version will be used.
        candidates : list
            The list of candidate versions to consider.
        strict_micro : bool
            If True, then ensure that the micro version matches exactly.
            If False (default), then let the micro version float to the newest version.
    
        Returns
        -------
        version : str
            The matching versions from the list of candidates.
    
        Raises
        ------
        NoMatchingVersions : if no matching version is found.
    
        Examples
        --------
        >>> find_version('3', ['3.1.0', '4.0.0', '4.1.0', '4.2.0-alpha0'])
        '3.1.0'
        >>> find_version('4', ['3.1.0', '4.0.0', '4.1.0', '4.2.0-alpha0'])
        '4.1.0'
        >>> find_version('4.0', ['3.1.0', '4.0.0', '4.1.0', '4.2.0-alpha0'])
        '4.0.0'
        >>> find_version(None, ['3.1.0', '4.0.0', '4.1.0', '4.2.0-alpha0'])
        '4.1.0'
        >>> find_version("4.2.0-alpha0", ['3.1.0', '4.0.0', '4.1.0', '4.2.0-alpha0'])
        '4.2.0-alpha0'
        """
        if not candidates:
            raise NoMatchingVersions("No candidate versions provided.")
        cand = sorted([Version(c) for c in candidates])
        if version is None:
            cand = sorted(c for c in cand if c.dev is None)
            if not candidates:
                raise NoMatchingVersions(f"No non-dev candidates in {candidates}.")
            return str(cand[-1])
    
        v = Version(version)
        if not strict_micro and not v.dev:
            v.micro = None
        matches = [c for c in cand if v.matches(c)]
        if not matches:
>           raise NoMatchingVersions(
                f"No matches for version={version!r} among {candidates}.\n"
                "Often this can be fixed by updating altair_viewer:\n"
                "    pip install -U altair_viewer"
            )
E           altair_viewer._utils.NoMatchingVersions: No matches for version='5.16.3' among ['4.0.2', '4.8.1', '4.17.0'].
E           Often this can be fixed by updating altair_viewer:
E               pip install -U altair_viewer

/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/altair_viewer/_utils.py:212: NoMatchingVersions
__________________________________________________________________________ test_save_svg_altair ___________________________________________________________________________

    def test_save_svg_altair():
        """Test that that figures saved as svg exist."""
        fig = alt.Chart(pd.DataFrame()).mark_bar()
        path = "tests/temp/"
>       saving.save(fig, "test_fig", path, save_png=False, save_svg=True)

tests/viz/altair/test_saving.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:115: in save
    driver = _google_chrome_driver_setup() if driver is None else driver
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:25: in _google_chrome_driver_setup
    driver = webdriver.Chrome(
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/selenium/webdriver/chrome/webdriver.py:70: in __init__
    super(WebDriver, self).__init__(DesiredCapabilities.CHROME['browserName'], "goog",
/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/selenium/webdriver/chromium/webdriver.py:89: in __init__
    self.service.start()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <selenium.webdriver.chrome.service.Service object at 0x16dfe3040>

    def start(self):
        """
        Starts the Service.
    
        :Exceptions:
         - WebDriverException : Raised either when it can't start the service
           or when it can't connect to the service
        """
        try:
            cmd = [self.path]
            cmd.extend(self.command_line_args())
            self.process = subprocess.Popen(cmd, env=self.env,
                                            close_fds=system() != 'Windows',
                                            stdout=self.log_file,
                                            stderr=self.log_file,
                                            stdin=PIPE,
                                            creationflags=self.creationflags)
        except TypeError:
            raise
        except OSError as err:
            if err.errno == errno.ENOENT:
                raise WebDriverException(
                    "'%s' executable needs to be in PATH. %s" % (
                        os.path.basename(self.path), self.start_error_message)
                )
            elif err.errno == errno.EACCES:
                raise WebDriverException(
                    "'%s' executable may have wrong permissions. %s" % (
                        os.path.basename(self.path), self.start_error_message)
                )
            else:
                raise
        except Exception as e:
            raise WebDriverException(
                "The executable %s needs to be available in the path. %s\n%s" %
                (os.path.basename(self.path), self.start_error_message, str(e)))
        count = 0
        while True:
            self.assert_process_still_running()
            if self.is_connectable():
                break
    
            count += 1
            sleep(0.5)
            if count == 60:
>               raise WebDriverException("Can not connect to the Service %s" % self.path)
E               selenium.common.exceptions.WebDriverException: Message: Can not connect to the Service /Users/solomonyu/.wdm/drivers/chromedriver/mac64/120.0.6099.129/chromedriver-mac-arm64/chromedriver

/opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/selenium/webdriver/common/service.py:105: WebDriverException
============================================================================ warnings summary =============================================================================
../../../../../opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/botocore/compat.py:31
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/botocore/compat.py:31: DeprecationWarning: 'urllib3[secure]' extra is deprecated and will be removed in a future release of urllib3 2.x. Read more in this issue: https://github.com/urllib3/urllib3/issues/2680
    from urllib3 import exceptions

tests/loading_saving/test_S3.py::test_upload_obj_unsup_data
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/loading_saving/S3.py:238: UserWarning: Data uploaded as pickle. Please consider other accessible file types among the supported ones.
    warnings.warn(

tests/loading_saving/test_S3.py::test_download_obj_dict_geojson
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/loading_saving/S3.py:392: UserWarning: Please check geojson has a member with the name 'type', the value of the member must be one of the following:'Point', 'MultiPoint', 'LineString', 'MultiLineString', 'Polygon', 'MultiPolygon', 'GeometryCollection','Feature' and 'FeatureCollection'. Else downloaded dictionary will not be valid geojson.
    warnings.warn(

tests/networks/test_build.py::test_network_as_adjacency_matrix
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/networks/build.py:87: FutureWarning: adjacency_matrix will return a scipy.sparse array instead of a matrix in Networkx 3.0.
    return nx.adjacency_matrix(network), nodes

tests/networks/test_build.py::test_network_as_adjacency_matrix
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/networkx/linalg/graphmatrix.py:173: DeprecationWarning: 
  
  The scipy.sparse array containers will be used instead of matrices
  in Networkx 3.0. Use `to_scipy_sparse_array` instead.
    return nx.to_scipy_sparse_matrix(G, nodelist=nodelist, dtype=dtype, weight=weight)

tests/viz/altair/test_format.py::test_nesta_theme_activation
tests/viz/altair/test_saving.py::test_save_png_altair
tests/viz/altair/test_saving.py::test_save_html_altair
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/formatting.py:25: UserWarning: Averta font could not be located. Using 'Helvetica' instead
    warnings.warn("Averta font could not be located. Using 'Helvetica' instead")

tests/viz/altair/test_saving.py::test_save_png_altair
tests/viz/altair/test_saving.py::test_save_svg_altair
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:25: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
    driver = webdriver.Chrome(

tests/viz/altair/test_saving.py::test_save_png_altair
tests/viz/altair/test_saving.py::test_save_svg_altair
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/nesta_ds_utils/viz/altair/saving.py:25: DeprecationWarning: use options instead of chrome_options
    driver = webdriver.Chrome(

tests/viz/altair/test_saving.py::test_save_png_altair
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/selenium/webdriver/remote/remote_connection.py:390: DeprecationWarning: HTTPResponse.getheader() is deprecated and will be removed in urllib3 v2.1.0. Instead use HTTPResponse.headers.get(name, default).
    if response.getheader('Content-Type'):

tests/viz/altair/test_saving.py::test_save_png_altair
  /opt/homebrew/Caskroom/miniconda/base/envs/ds_utils/lib/python3.10/site-packages/selenium/webdriver/remote/remote_connection.py:391: DeprecationWarning: HTTPResponse.getheader() is deprecated and will be removed in urllib3 v2.1.0. Instead use HTTPResponse.headers.get(name, default).
    content_type = response.getheader('Content-Type').split(';')

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================= short test summary info =========================================================================
FAILED tests/viz/altair/test_saving.py::test_save_png_altair - altair_viewer._utils.NoMatchingVersions: No matches for version='5.16.3' among ['4.0.2', '4.8.1', '4.17.0'].
FAILED tests/viz/altair/test_saving.py::test_save_svg_altair - selenium.common.exceptions.WebDriverException: Message: Can not connect to the Service /Users/solomonyu/....
=============================================================== 2 failed, 62 passed, 14 warnings in 36.89s ================================================================
```

--- a/docs/source/saving.rst
+++ b/docs/source/saving.rst
@@ -2,4 +2,5 @@ saving
 ===============
 **saving** contains altair saving functions
 
+.. autofunction:: nesta_ds_utils.viz.altair.saving.webdriver_context
 .. autofunction:: nesta_ds_utils.viz.altair.saving.save

--- a/nesta_ds_utils/loading_saving/S3.py
+++ b/nesta_ds_utils/loading_saving/S3.py
@@ -3,7 +3,6 @@ from typing import List
 import boto3
 from fnmatch import fnmatch
 import pandas as pd
-import geopandas as gpd
 import numpy as np
 import pyarrow as pa
 import pyarrow.parquet as pq
@@ -11,6 +10,14 @@ import json
 import pickle
 import warnings
 from nesta_ds_utils.loading_saving import file_ops
+
+from nesta_ds_utils.loading_saving.gis_interface import _gis_enabled
+
+if _gis_enabled:
+    from nesta_ds_utils.loading_saving.gis_interface import (
+        _gdf_to_fileobj,
+        _fileobj_to_gdf,
+    )
 
 
 def get_bucket_filenames_s3(bucket_name: str, dir_name: str = "") -> List[str]:
@@ -53,27 +60,6 @@ def _df_to_fileobj(df_data: pd.DataFrame, path_to: str, **kwargs) -> io.BytesIO:
     else:
         raise NotImplementedError(
             "Uploading dataframe currently supported only for 'csv', 'parquet', 'xlsx' and xlsm'."
-        )
-    buffer.seek(0)
-    return buffer
-
-
-def _gdf_to_fileobj(df_data: gpd.GeoDataFrame, path_to: str, **kwargs) -> io.BytesIO:
-    """Convert GeoDataFrame into bytes file object.
-
-    Args:
-        df_data (gpd.DataFrame): Dataframe to convert.
-        path_to (str): Saving file name.
-
-    Returns:
-        io.BytesIO: Bytes file object.
-    """
-    buffer = io.BytesIO()
-    if fnmatch(path_to, "*.geojson"):
-        df_data.to_file(buffer, driver="GeoJSON", **kwargs)
-    else:
-        raise NotImplementedError(
-            "Uploading geodataframe currently supported only for 'geojson'."
         )
     buffer.seek(0)
     return buffer
@@ -229,10 +215,16 @@ def upload_obj(
         kwargs_writing (dict, optional): Dictionary of kwargs for writing data.
 
     """
-    if isinstance(obj, gpd.base.GeoPandasBase):
-        obj = _gdf_to_fileobj(obj, path_to, **kwargs_writing)
-    elif isinstance(obj, pd.DataFrame):
-        obj = _df_to_fileobj(obj, path_to, **kwargs_writing)
+    if isinstance(obj, pd.DataFrame):
+        if type(obj).__name__ == "GeoDataFrame":
+            if _gis_enabled:
+                obj = _gdf_to_fileobj(obj, path_to, **kwargs_writing)
+            else:
+                raise ModuleNotFoundError(
+                    "Please install 'gis' extra from nesta_ds_utils or 'geopandas' to upload geodataframes."
+                )
+        else:
+            obj = _df_to_fileobj(obj, path_to, **kwargs_writing)
     elif isinstance(obj, dict):
         obj = _dict_to_fileobj(obj, path_to, **kwargs_writing)
     elif isinstance(obj, list):
@@ -270,22 +262,6 @@ def _fileobj_to_df(fileobj: io.BytesIO, path_from: str, **kwargs) -> pd.DataFram
         return pd.read_excel(fileobj, **kwargs)
     elif fnmatch(path_from, "*.xlsm"):
         return pd.read_excel(fileobj, **kwargs)
-
-
-def _fileobj_to_gdf(fileobj: io.BytesIO, path_from: str, **kwargs) -> pd.DataFrame:
-    """Convert bytes file object into geodataframe.
-
-    Args:
-        fileobj (io.BytesIO): Bytes file object.
-        path_from (str): Path of loaded data.
-
-    Returns:
-        gpd.DataFrame: Data as geodataframe.
-    """
-    if fnmatch(path_from, "*.geojson"):
-        return gpd.GeoDataFrame.from_features(
-            json.loads(fileobj.getvalue().decode())["features"]
-        )
 
 
 def _fileobj_to_dict(fileobj: io.BytesIO, path_from: str, **kwargs) -> dict:
@@ -399,7 +375,12 @@ def download_obj(
             )
     elif download_as == "geodf":
         if path_from.endswith(tuple([".geojson"])):
-            return _fileobj_to_gdf(fileobj, path_from, **kwargs_reading)
+            if _gis_enabled:
+                return _fileobj_to_gdf(fileobj, path_from, **kwargs_reading)
+            else:
+                raise ModuleNotFoundError(
+                    "Please install 'gis' extra from nesta_ds_utils or 'geopandas' to download geodataframes."
+                )
         else:
             raise NotImplementedError(
                 "Download as geodataframe currently supported only " "for 'geojson'."

--- a/nesta_ds_utils/loading_saving/gis_interface.py
+++ b/nesta_ds_utils/loading_saving/gis_interface.py
@@ -1,0 +1,45 @@
+try:
+    from json import loads as load_json
+    from geopandas import GeoDataFrame
+    from io import BytesIO
+    from fnmatch import fnmatch
+
+    _gis_enabled = True
+
+    def _gdf_to_fileobj(df_data: GeoDataFrame, path_to: str, **kwargs) -> BytesIO:
+        """Convert GeoDataFrame into bytes file object.
+
+        Args:
+            df_data (gpd.DataFrame): Dataframe to convert.
+            path_to (str): Saving file name.
+
+        Returns:
+            io.BytesIO: Bytes file object.
+        """
+        buffer = BytesIO()
+        if fnmatch(path_to, "*.geojson"):
+            df_data.to_file(buffer, driver="GeoJSON", **kwargs)
+        else:
+            raise NotImplementedError(
+                "Uploading geodataframe currently supported only for 'geojson'."
+            )
+        buffer.seek(0)
+        return buffer
+
+    def _fileobj_to_gdf(fileobj: BytesIO, path_from: str, **kwargs) -> GeoDataFrame:
+        """Convert bytes file object into geodataframe.
+
+        Args:
+            fileobj (io.BytesIO): Bytes file object.
+            path_from (str): Path of loaded data.
+
+        Returns:
+            gpd.DataFrame: Data as geodataframe.
+        """
+        if fnmatch(path_from, "*.geojson"):
+            return GeoDataFrame.from_features(
+                load_json(fileobj.getvalue().decode())["features"]
+            )
+
+except ImportError:
+    _gis_enabled = False

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -99,13 +99,13 @@ def save(
         save_svg (bool, optional): Option to save figure as 'svg'. Default to False.
         scale_factor (int, optional): Saving scale factor. Default to 5.
     """
-    path = file_ops._convert_str_to_pathlib_path(path)
-
     if not any([save_png, save_html, save_svg]):
         raise Exception(
             "At least one format needs to be selected. Example: save(.., save_png=True)."
         )
 
+    path = file_ops._convert_str_to_pathlib_path(path)
+    file_ops.make_path_if_not_exist(path)
     if save_png or save_svg:
         driver = _google_chrome_driver_setup() if driver is None else driver
 

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -14,6 +14,7 @@ from matplotlib import font_manager
 from pathlib import Path
 from nesta_ds_utils.loading_saving import file_ops
 import yaml
+from contextlib import contextmanager
 
 
 def _google_chrome_driver_setup() -> WebDriver:
@@ -23,6 +24,15 @@ def _google_chrome_driver_setup() -> WebDriver:
     chrome_options.add_argument("--headless=new")
     driver = Chrome(service=service, options=chrome_options)
     return driver
+
+
+@contextmanager
+def webdriver_context(driver: WebDriver = None):
+    try:
+        driver = _google_chrome_driver_setup() if driver is None else driver
+        yield driver
+    finally:
+        driver.quit()
 
 
 def _save_png(

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -116,19 +116,18 @@ def save(
 
     path = file_ops._convert_str_to_pathlib_path(path)
     file_ops.make_path_if_not_exist(path)
-    if save_png or save_svg:
-        driver = _google_chrome_driver_setup() if driver is None else driver
 
-    file_ops.make_path_if_not_exist(path)
-    # Export figures
-    if save_png:
-        _save_png(fig, path, name, scale_factor, driver)
+    if save_png or save_svg:
+        with webdriver_context(driver):
+            # Export figures
+            if save_png:
+                _save_png(fig, path, name, scale_factor, driver)
+
+            if save_svg:
+                _save_svg(fig, path, name, scale_factor, driver)
 
     if save_html:
         _save_html(fig, path, name, scale_factor)
-
-    if save_svg:
-        _save_svg(fig, path, name, scale_factor, driver)
 
 
 def _find_averta() -> str:

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -5,10 +5,9 @@ Module containing utils for styling and exporting figures using Altair.
 import altair_saver as alt_saver
 from altair.vegalite import Chart
 import altair as alt
-from selenium import webdriver
 from webdriver_manager.chrome import ChromeDriverManager
 from selenium.webdriver.chrome.webdriver import WebDriver
-from selenium.webdriver.chrome.options import Options
+from selenium.webdriver import Chrome, ChromeOptions, ChromeService
 import os
 from typing import Union, List, Type
 import warnings
@@ -20,11 +19,10 @@ import yaml
 
 def _google_chrome_driver_setup() -> WebDriver:
     """Set up the driver to save figures"""
-    chrome_options = Options()
-    chrome_options.add_argument("--headless")
-    driver = webdriver.Chrome(
-        ChromeDriverManager().install(), chrome_options=chrome_options
-    )
+    service = ChromeService(ChromeDriverManager().install())
+    chrome_options = ChromeOptions()
+    chrome_options.add_argument("--headless=new")
+    driver = Chrome(service=service, options=chrome_options)
     return driver
 
 

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -2,7 +2,6 @@
 Module containing utils for styling and exporting figures using Altair.
 """
 
-import altair_saver as alt_saver
 from altair.vegalite import Chart
 import altair as alt
 from webdriver_manager.chrome import ChromeDriverManager
@@ -38,8 +37,7 @@ def _save_png(
         scale_factor (int): Saving scale factor.
         driver (WebDriver): webdriver to use for saving.
     """
-    alt_saver.save(
-        fig,
+    fig.save(
         f"{path}/{name}.png",
         method="selenium",
         webdriver=driver,
@@ -71,8 +69,7 @@ def _save_svg(
         scale_factor (int): Saving scale factor.
         driver (WebDriver): webdriver to use for saving.
     """
-    alt_saver.save(
-        fig,
+    fig.save(
         f"{path}/{name}.svg",
         method="selenium",
         scale_factor=scale_factor,

--- a/nesta_ds_utils/viz/altair/saving.py
+++ b/nesta_ds_utils/viz/altair/saving.py
@@ -28,6 +28,21 @@ def _google_chrome_driver_setup() -> WebDriver:
 
 @contextmanager
 def webdriver_context(driver: WebDriver = None):
+    """Context Manager for Selenium WebDrivers.
+    Optionally pass in user-instantiated Selenium Webdriver.
+    Defaults to setup and yield a ChromeWebDriver.
+
+    Typical usage:
+
+        with webdriver_context(webdriver or None) as driver:
+            # Do stuff with driver, driver.quit() is then called automatically
+
+    Args:
+        driver (WebDriver, optional): Webdriver to use. Defaults to 'webdriver.Chrome'.
+
+    Yields:
+        WebDriver: The optional user-instantiated Selenium WebDriver or a Selenium ChromeWebDriver.
+    """
     try:
         driver = _google_chrome_driver_setup() if driver is None else driver
         yield driver

--- a/setup.cfg
+++ b/setup.cfg
@@ -12,37 +12,39 @@ packages = ["nesta_ds_utils"]
 [options]
 python_requires = >=3.8
 install_requires =
-    numpy==1.23.4
-    pandas==1.5.1
+    numpy>=1.23.4
+    pandas>=1.5.1
     pyyaml<5.4.0
-    scipy==1.9.3
-    pyarrow==10.0.0
+    scipy>=1.9.3
+    pyarrow>=10.0.0
 [options.extras_require]
 s3 =
-  boto3==1.24.93
+  boto3>=1.24.93
 gis =
-  geopandas==0.13.2
-io_extras = 
-  openpyxl==3.0.9
+  geopandas>=0.13.2
+io_extras =
+  openpyxl>=3.0.9
 viz =
-  altair==4.2.0
-  altair-saver==0.5.0
-  matplotlib==3.6.2
-  selenium==4.2.0
-  webdriver_manager==4.0.0
+  altair>=4.2.0
+  vl-convert-python>=1.2.0
+  matplotlib>=3.6.2
+  selenium>=4.2.0
+  webdriver_manager>=4.0.0
 networks =
   networkx==2.8.8
 nlp =
-  nltk==3.7
-test =
-  pytest==7.1.3
-  moto[s3]==4.0.7
+  nltk>=3.7
+all =
   %(s3)s
   %(gis)s
   %(io_extras)s
   %(viz)s
   %(networks)s
   %(nlp)s
+test =
+  pytest==7.1.3
+  moto[s3]==4.0.7
+  %(all)s
 dev =
   Sphinx==5.2.3
   sphinxcontrib-applehelp==1.0.2
@@ -51,24 +53,10 @@ dev =
   sphinxcontrib-jsmath==1.0.1
   sphinxcontrib-qthelp==1.0.3
   sphinxcontrib-serializinghtml==1.1.5
-  pytest==7.1.3
-  moto[s3]==4.0.7
   pre-commit==2.20.0
   pre-commit-hooks==4.3.0
   black==22.10.0
-  %(s3)s
-  %(gis)s
-  %(io_extras)s
-  %(viz)s
-  %(networks)s
-  %(nlp)s
-all =
-  %(s3)s
-  %(gis)s
-  %(io_extras)s
-  %(viz)s
-  %(networks)s
-  %(nlp)s
+  %(test)s
 
 [options.package_data]
 nesta_ds_utils.viz.themes =

--- a/tests/viz/altair/test_saving.py
+++ b/tests/viz/altair/test_saving.py
@@ -1,6 +1,7 @@
 import shutil
 from pathlib import Path
 from nesta_ds_utils.viz.altair import saving
+from selenium.webdriver.chromium.webdriver import ChromiumDriver
 import pandas as pd
 import altair as alt
 import pytest
@@ -41,3 +42,18 @@ def test_save_altair_exception():
         saving.save(
             fig, "test_fig", path, save_png=False, save_html=False, save_svg=False
         )
+
+
+def test_webdriver():
+    """Test that Chrome WebDriver is created by default is a ChromiumDriver, and context manager stops the webdriver."""
+    driver = saving._google_chrome_driver_setup()
+    assert isinstance(driver, ChromiumDriver)
+
+    with saving.webdriver_context(driver) as some_driver:
+        # No actions needed here,
+        # just testing that the context manager calls .quit() on driver to terminate.
+        pass
+
+    # If subprocess not terminated, .poll() returns None
+    # https://docs.python.org/3/library/subprocess.html#subprocess.Popen.returncode
+    assert driver.service.process.poll() is not None

